mungegithub/mungers/publish_scripts/util.sh

```diff
@@ -19,118 +19,551 @@
-# 5. switch back to the "target-branch"
-# 6. for each commit, cherry-pick it (which will keep authorship) into "target-branch"
-# 7. update metadata files indicating which commits we've sync'ed to
+# 2. we are on the branch to be published (let's call it "destination branch"), possibly this
+#    is an orphaned branch for new branches.
+#
+# Overall Flow
+# ============
+#
+# 1. Fetch the current level of k8s.io/kubernetes.
+# 2. Check out the ${src_branch} of k8s.io/kubernetes as branch filter-branch.
+# 3. Rewrite the history of branch filter-branch to *only* include code in ${subdirectory},
+#    keeping the original, corresponding k8s.io/kubernetes via a "Kubernetes-commit: <kube-sha>"
+#    line in the commit message.
+# 4. Locate all commits between the last time we sync'ed and now on the mainline, i.e. using
+#    first-parent ancestors only, not going into a PR branch.
+# 5. Switch back to the ${dst_branch}
+# 6. For each commit C on the mainline (= ${f_mainline_commit} in the code) identified in 4:
+#    a) If C is a merge commit:
+#       i)   Find the latest branching point B from the mainline leading to C as its second parent.
+#       ii)  If there is another merge between B and C via the path of (i), apply the diff between
+#            the latest branching point B and the last merge M on that path.
+#       iii) Continue cherry-picking all single commits between B (if (ii) didn’t apply) or M
+#            (by (ii) these are no merges).
+#       iv)  Commit the merge commit C as empty fast-forward merge of the latest branching point
+#            and the HEAD produced from of (ii) and (iii).
+#    b) If C is not a merge commit:
+#       i)   Find the corresponding k8s.io/kubernetes commit k_C and find its merge point k_M
+#            with the mainline in k8s.io/kubernetes.
+#       ii)  Cherry-pick k_C.
+#       iii) Continue with the next C in 6, let’s call it C’. If
+#            - there is no such C’  or
+#            - C’ is a merge commit or
+#            - the mainline merge point k_M’ of the corresponding k8s.io/kubernetes commit k_C’
+#              is different from k_M,
+#            cherry-pick k_M as fast-forward merge commit of the mainline and C.
+#
+# Dropped PR Merges
+# =================
+#
+# The logic of 6b is necessary because git filter-branch drops fast-forward merges, leading to
+# a picture like in the next section..
+#
+# With 6b we get a perfectly uniform sequence of branches and fast-forward merges, i.e. without
+# any interleaving. For this to work it is essential that Github’s merge commit are clean as
+# described above.
+#
+# Result
+# ======
+#
+# This gives the following picture. All merges on the mainline are fast-forward merges.
+# All branches of the second parents are linear.
+#
+#   M─┐ [master] Merge pull request #51467 from liggitt/client-go-owner
+#   │ o Add liggitt to client-go approvers
+#   │─┘
+#   I─┐ Merge pull request #50562 from atlassian/call-cleanup-properly
```

Comment:
What's the difference between M and I?

Reply:
that was tig output. Probably `I` is fast-forward.

Reply:
Changed to M